File: src/main/java/vini2003/xyz/bambootweaks/registry/BambooTweaksBlocks.java
package vini2003.xyz.bambootweaks.registry;

import net.fabricmc.fabric.api.object.builder.v1.block.FabricBlockSettings;
import vini2003.xyz.bambootweaks.BambooTweaks;
import vini2003.xyz.bambootweaks.block.*;
import net.fabricmc.fabric.impl.content.registry.FuelRegistryImpl;
import net.minecraft.block.Block;
import net.minecraft.block.FenceBlock;
import net.minecraft.block.FenceGateBlock;
import net.minecraft.block.Material;
import net.minecraft.item.BlockItem;
import net.minecraft.item.Item;
import net.minecraft.item.ItemGroup;
import net.minecraft.particle.ParticleTypes;
import net.minecraft.sound.BlockSoundGroup;
import net.minecraft.util.Identifier;
import net.minecraft.util.registry.Registry;

public class BambooTweaksBlocks {
	public static final Block BAMBOO_BLOCK = register("bamboo_block", new Block(FabricBlockSettings.of(Material.BAMBOO).sounds(BlockSoundGroup.BAMBOO).strength(0.5F, 2.5F)), new Item.Settings().group(BambooTweaksItemGroups.BAMBOO_TWEAKS));
	public static final Block BAMBOO_DOOR = register("bamboo_door", new ExposedDoorBlock(FabricBlockSettings.of(Material.BAMBOO).strength(3.0F, 0.0F).sounds(BlockSoundGroup.BAMBOO).nonOpaque()), new Item.Settings().group(BambooTweaksItemGroups.BAMBOO_TWEAKS));
	public static final Block BAMBOO_FENCE_BLOCK = register("bamboo_fence", new FenceBlock(FabricBlockSettings.of(Material.BAMBOO).sounds(BlockSoundGroup.BAMBOO).strength(2.0F, 3.0F)), new Item.Settings().group(BambooTweaksItemGroups.BAMBOO_TWEAKS));
	public static final Block BAMBOO_FENCE_GATE_BLOCK = register("bamboo_fence_gate", new FenceGateBlock(FabricBlockSettings.of(Material.BAMBOO).sounds(BlockSoundGroup.BAMBOO).strength(2.0F, 3.0F)), new Item.Settings().group(BambooTweaksItemGroups.BAMBOO_TWEAKS));
	public static final Block BAMBOO_FENCE_WALL_BLOCK = register("bamboo_wall", new BlockBambooWall(FabricBlockSettings.of(Material.BAMBOO).sounds(BlockSoundGroup.BAMBOO).strength(2.0F, 3.0F)), new Item.Settings().group(BambooTweaksItemGroups.BAMBOO_TWEAKS));
	public static final Block BAMBOO_LADDER_BLOCK = register("bamboo_ladder", new ExposedLadderBlock(FabricBlockSettings.of(Material.BAMBOO).sounds(BlockSoundGroup.BAMBOO).strength(0.4F, 0.0F).nonOpaque()), new Item.Settings().group(BambooTweaksItemGroups.BAMBOO_TWEAKS));
	public static final Block BAMBOO_ROD = register("bamboo_rod", new BlockBambooRod(FabricBlockSettings.of(Material.BAMBOO).sounds(BlockSoundGroup.BAMBOO).breakInstantly()), new Item.Settings().group(BambooTweaksItemGroups.BAMBOO_TWEAKS));
	public static final Block BAMBOO_SPIKE_BLOCK = register("bamboo_spike", new BlockBambooSpike(FabricBlockSettings.of(Material.BAMBOO).sounds(BlockSoundGroup.BAMBOO).strength(2.0F, 3.0F)), new Item.Settings().group(BambooTweaksItemGroups.BAMBOO_TWEAKS));
	public static final Block BAMBOO_STAIRS_BLOCK = register("bamboo_stairs", new ExposedStairsBlock(BAMBOO_BLOCK.getDefaultState(), FabricBlockSettings.of(Material.BAMBOO).sounds(BlockSoundGroup.BAMBOO).strength(2.0F, 3.0F)), new Item.Settings().group(BambooTweaksItemGroups.BAMBOO_TWEAKS));
	public static final Block BAMBOO_TORCH_BLOCK = register("bamboo_torch", new ExposedTorchBlock(FabricBlockSettings.of(Material.ORGANIC_PRODUCT).luminance(14).noCollision().sounds(BlockSoundGroup.BAMBOO).nonOpaque(), ParticleTypes.FLAME), (BlockItem) null);
	public static final Block BAMBOO_TORCH_BLOCK_WALL = register("bamboo_torch_wall", new ExposedWallTorchBlock(FabricBlockSettings.of(Material.ORGANIC_PRODUCT).luminance(14).noCollision().sounds(BlockSoundGroup.BAMBOO).nonOpaque(), ParticleTypes.FLAME), (BlockItem) null);

	public BambooTweaksBlocks() {
		// NO-OP
	}

	public static void initialize() {
		FuelRegistryImpl.INSTANCE.add(BAMBOO_BLOCK, 500);
		FuelRegistryImpl.INSTANCE.add(BAMBOO_DOOR, 200);
		FuelRegistryImpl.INSTANCE.add(BAMBOO_FENCE_BLOCK, 300);
		FuelRegistryImpl.INSTANCE.add(BAMBOO_FENCE_GATE_BLOCK, 300);
		FuelRegistryImpl.INSTANCE.add(BAMBOO_FENCE_WALL_BLOCK, 300);
		FuelRegistryImpl.INSTANCE.add(BAMBOO_LADDER_BLOCK, 300);
		FuelRegistryImpl.INSTANCE.add(BAMBOO_ROD, 200);
		FuelRegistryImpl.INSTANCE.add(BAMBOO_STAIRS_BLOCK, 300);
		FuelRegistryImpl.INSTANCE.add(BAMBOO_SPIKE_BLOCK, 300);
	}

	static <T extends Block> T register(String name, T block, Item.Settings settings) {
		return register(name, block, new BlockItem(block, settings));
	}

	static <T extends Block> T register(String name, T block, BlockItem item) {
		T b = Registry.register(Registry.BLOCK, new Identifier(BambooTweaks.ID, name), block);
		if (item != null) {
			BambooTweaksItems.register(name, item);
		}
		return b;
	}
}

File: common/src/main/kotlin/vini2003/xyz/bambootweaks/common/registry/BambooTweaksBlocks.kt
package vini2003.xyz.bambootweaks.common.registry

import me.shedaniel.architectury.registry.DeferredRegister
import me.shedaniel.architectury.registry.RegistrySupplier
import net.minecraft.block.*
import net.minecraft.item.BlockItem
import net.minecraft.item.Item
import net.minecraft.particle.ParticleTypes
import net.minecraft.sound.BlockSoundGroup
import net.minecraft.util.registry.Registry
import vini2003.xyz.bambootweaks.CommonBambooTweaks
import vini2003.xyz.bambootweaks.common.block.*

object BambooTweaksBlocks {
	private val Blocks = BambooTweaksDeferredRegisters.Blocks
	private val Items = BambooTweaksDeferredRegisters.Items
	
	fun init() {

	}
	
	fun <T : Block> register(name: String, supplier: () -> T): RegistrySupplier<T> {
		return Blocks.register(name, supplier)
	}
	
	fun <T : Block> register(name: String, supplier: () -> T, settings: (Item.Settings) -> (Item.Settings)): RegistrySupplier<T> {
		Blocks.register(name, supplier).let { block ->
			Items.register(name) {
				BlockItem(block.get(), settings.invoke(itemSettings()))
			}
			
			return block
		}
	}
	
	private fun blockSettings(material: Material) = AbstractBlock.Settings.of(material)!!
	
	private fun itemSettings() = Item.Settings().group(BambooTweaksItemGroups.BambooTweaks)!!
	
	val Bamboo = register("bamboo_block", {
		Block(blockSettings(Material.BAMBOO).strength(2.0F, 3.0F))
	}, { it })
	
	val BambooDoor = register("bamboo_door", {
		ExposedDoorBlock(blockSettings(Material.BAMBOO).strength(3.0F).nonOpaque())
	}, { it })
	
	val BambooFence = register("bamboo_fence", {
		FenceBlock(blockSettings(Material.BAMBOO).strength(2.0F, 3.0F))
	}, { it })
	
	val BambooFenceGate = register("bamboo_fence_gate", {
		FenceGateBlock(blockSettings(Material.BAMBOO).strength(2.0F, 3.0F))
	}, { it })
	
	val BambooWall = register("bamboo_wall", {
		BambooWallBlock(blockSettings(Material.BAMBOO).strength(2.0F, 3.0F))
	}, { it })
	
	val BambooLadder = register("bamboo_ladder", {
		ExposedLadderBlock(blockSettings(Material.BAMBOO).sounds(BlockSoundGroup.LADDER).strength(0.4F).nonOpaque())
	}, { it })
	
	val BambooRod = register("bamboo_rod", {
		BambooRodBlock(blockSettings(Material.SUPPORTED).breakInstantly().nonOpaque())
	}, { it })
	
	val BambooSpikes = register("bamboo_spikes", {
		BambooSpikesBlock(blockSettings(Material.SUPPORTED).breakInstantly().nonOpaque())
	}, { it })
	
	val BambooStairs = register("bamboo_stairs", {
		ExposedStairsBlock(Bamboo.orNull!!.defaultState, blockSettings(Material.BAMBOO).strength(2.0F, 3.0F))
	}, { it })
	
	val BambooTorch = register("bamboo_torch") {
		ExposedTorchBlock(blockSettings(Material.SUPPORTED).noCollision().breakInstantly().luminance { 14 }.noCollision(), ParticleTypes.FLAME)
	}
	
	val BambooWallTorch = register("bamboo_wall_torch") {
		ExposedWallTorchBlock(blockSettings(Material.SUPPORTED).noCollision().breakInstantly().luminance { 14 }.noCollision(), ParticleTypes.FLAME)
	}
	
	val DryBamboo = register("dry_bamboo_block", {
		Block(blockSettings(Material.BAMBOO).strength(2.0F, 3.0F))
	}, { it })
	
	val DryBambooDoor = register("dry_bamboo_door", {
		ExposedDoorBlock(blockSettings(Material.BAMBOO).strength(3.0F).nonOpaque())
	}, { it })
	
	val DryBambooFence = register("dry_bamboo_fence", {
		FenceBlock(blockSettings(Material.BAMBOO).strength(2.0F, 3.0F))
	}, { it })
	
	val DryBambooFenceGate = register("dry_bamboo_fence_gate", {
		FenceGateBlock(blockSettings(Material.BAMBOO).strength(2.0F, 3.0F))
	}, { it })
	
	val DryBambooWall = register("dry_bamboo_wall", {
		BambooWallBlock(blockSettings(Material.BAMBOO).strength(2.0F, 3.0F))
	}, { it })
	
	val DryBambooLadder = register("dry_bamboo_ladder", {
		ExposedLadderBlock(blockSettings(Material.BAMBOO).sounds(BlockSoundGroup.LADDER).strength(0.4F).nonOpaque())
	}, { it })
	
	val DryBambooRod = register("dry_bamboo_rod", {
		BambooRodBlock(blockSettings(Material.SUPPORTED).breakInstantly().nonOpaque())
	}, { it })
	
	val DryBambooSpikes = register("dry_bamboo_spikes", {
		BambooSpikesBlock(blockSettings(Material.SUPPORTED).breakInstantly().nonOpaque())
	}, { it })
	
	val DryBambooStairs = register("dry_bamboo_stairs", {
		ExposedStairsBlock(DryBamboo.orNull!!.defaultState, blockSettings(Material.BAMBOO).strength(2.0F, 3.0F))
	}, { it })
	
	val DryBambooTorch = register("dry_bamboo_torch") {
		ExposedTorchBlock(blockSettings(Material.SUPPORTED).noCollision().breakInstantly().luminance { 14 }.noCollision(), ParticleTypes.FLAME)
	}
	
	val DryBambooWallTorch = register("dry_bamboo_wall_torch") {
		ExposedWallTorchBlock(blockSettings(Material.SUPPORTED).noCollision().breakInstantly().luminance { 14 }.noCollision(), ParticleTypes.FLAME)
	}
}